src/solvers/dgsem_tree/subcell_limiters_2d.jl

```diff
@@ -291,4 +321,180 @@ end
+    for iter in 1:(limiter.max_iterations_newton)
+        beta_old = beta
+
+        # If the state is valid, evaluate d(goal)/d(beta)
+        if is_valid_state(u_curr, equations)
+            dgoal_dbeta = dgoal_function(variable, u_curr, dt, antidiffusive_flux,
+                                         equations)
+        else # Otherwise, perform a bisection step
+            dgoal_dbeta = 0
+        end
+
+        if dgoal_dbeta != 0
+            # Update beta with Newton's method
+            beta = beta - goal / dgoal_dbeta
+        end
+
+        # Check bounds
+        if (beta < beta_L) || (beta > beta_R) || (dgoal_dbeta == 0) || isnan(beta)
+            # Out of bounds, do a bisection step
+            beta = 0.5 * (beta_L + beta_R)
+            # Get new u
+            u_curr = u + beta * dt * antidiffusive_flux
+
+            # If the state is invalid, finish bisection step without checking tolerance and iterate further
+            if !is_valid_state(u_curr, equations)
+                beta_R = beta
+                continue
+            end
+
+            # Check new beta for condition and update bounds
+            goal = goal_function(variable, bound, u_curr, equations)
+            if initial_check(bound, goal, newton_abstol)
+                # New beta fulfills condition
+                beta_L = beta
+            else
+                # New beta does not fulfill condition
+                beta_R = beta
+            end
+        else
+            # Get new u
+            u_curr = u + beta * dt * antidiffusive_flux
+
+            # If the state is invalid, redefine right bound without checking tolerance and iterate further
+            if !is_valid_state(u_curr, equations)
+                beta_R = beta
+                continue
+            end
+
+            # Evaluate goal function
+            goal = goal_function(variable, bound, u_curr, equations)
+        end
+
+        # Check relative tolerance
+        if abs(beta_old - beta) <= newton_reltol
+            break
+        end
+
+        # Check absolute tolerance
+        if final_check(bound, goal, newton_abstol)
+            break
+        end
+    end
+
+    new_alpha = 1 - beta
+    if alpha[i, j, element] > new_alpha + newton_abstol
+        error("Alpha is getting smaller. old: $(alpha[i, j, element]), new: $new_alpha")
+    else
+        alpha[i, j, element] = new_alpha
+    end
+
+    return nothing
+end
+
+# Initial checks
+@inline initial_check_nonnegative(bound, goal, newton_abstol) = goal <= 0
+
+# Goal and d(Goal)d(u) function
+@inline goal_function(variable, bound, u, equations) = bound - variable(u, equations)
+@inline function dgoal_function(variable, u, dt, antidiffusive_flux, equations)
+    -dot(variable_derivative(variable, u, equations), dt * antidiffusive_flux)
+end
+
+# Final checks
+@inline function final_check_nonnegative(bound, goal, newton_abstol)
+    (goal <= eps()) && (goal > -max(newton_abstol, abs(bound) * newton_abstol))
+end
```

Comment:
I can't claim I know what these functions are doing, but could they maybe be renamed to something less generic? "goal_function" sounds very general but is used very specifically here - maybe `goal_function_idp`?

Reply:
Done in [b776b6f](https://github.com/trixi-framework/Trixi.jl/pull/1738/commits/b776b6f226a05ec2bb6fdcfa5ef10f74c91093f9).